invokeai/app/invocations/compel.py

```diff
@@ -0,0 +1,283 @@
+from typing import Literal, Optional, Union
+from pydantic import BaseModel, Field
+
+from invokeai.app.invocations.util.choose_model import choose_model
+from .baseinvocation import BaseInvocation, BaseInvocationOutput, InvocationContext, InvocationConfig
+
+from ...backend.util.devices import choose_torch_device, torch_dtype
+from ...backend.stable_diffusion.diffusion import InvokeAIDiffuserComponent
+from ...backend.stable_diffusion.textual_inversion_manager import TextualInversionManager
+
+from compel import Compel
+from compel.prompt_parser import (
+    Blend,
+    CrossAttentionControlSubstitute,
+    FlattenedPrompt,
+    Fragment,
+)
+
+from invokeai.backend.globals import Globals
+
+
+class ConditioningField(BaseModel):
+    conditioning_name: Optional[str] = Field(default=None, description="The name of conditioning data")
+    class Config:
+        schema_extra = {"required": ["conditioning_name"]}
+
+
+class CompelOutput(BaseInvocationOutput):
+    """Compel parser output"""
+
+    #fmt: off
+    type: Literal["compel_output"] = "compel_output"
+    # name + loras -> pipeline + loras
+    # model: ModelField           = Field(default=None, description="Model")
+    # src? + loras -> tokenizer + text_encoder + loras
+    # clip:  ClipField            = Field(default=None, description="Text encoder(clip)")
+    positive: ConditioningField = Field(default=None, description="Positive conditioning")
+    negative: ConditioningField = Field(default=None, description="Negative conditioning")
+    #fmt: on
+
+
+class CompelInvocation(BaseInvocation):
+
+    type: Literal["compel"] = "compel"
+
+    positive_prompt: str = Field(default="", description="Positive prompt")
```

Comment:
Kind of same comment as above - it'd potentially be nice if this just generated a single set of conditioning (positive or negative). Then the parameter could just be "prompt".

Also, does compel not process negative prompts using brackets anymore? I haven't been using it much in a while - has our prompt syntax changed?

Reply:
Compel just translates prompt string to conditioning tensor, it's don't know is this tensor from positive prompt or from negative. It's invoke parses [] now.
So, if do more precisely we need to do compel with only 1 field and user need to create 2 compel nodes: first - for positive prompt parsing and second for negative prompt parsing.
I also firstly thinks about only 1 prompt input and output(so user will create 2 compel nodes), but now we not sure how loras will work and if loras will be applied in compel from prompts it's better to be 2 inputs in node as we can easily parse them with same text_encoder state(and that also why in my comments compel has output unet and clip(text_encoder)).